File: apps/ingestor/internal/event/pipeline_test.go
package event

import (
	"net/http"
	"strings"
	"testing"
)

const chromeUA = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 " +
	"(KHTML, like Gecko) Chrome/120 Safari/537.36"

func testBeacon() *Beacon {
	return &Beacon{
		N: "pageview",
		U: "https://example.com/pricing?utm_source=hn",
		D: "example.com",
		R: "https://news.ycombinator.com/",
	}
}

func testCtx(salt string) RequestContext {
	h := http.Header{}
	h.Set("X-Forwarded-For", "203.0.113.7")
	h.Set("Cf-Ipcountry", "DE")
	return RequestContext{Header: h, UserAgent: chromeUA, Salt: salt, SiteID: 42}
}

func TestBuildRow_DerivedColumns(t *testing.T) {
	row := BuildRow(testBeacon(), testCtx("day-1"))
	checks := map[string]struct{ got, want string }{
		"hostname":   {row.Hostname, "example.com"},
		"pathname":   {row.Pathname, "/pricing"},
		"referrer":   {row.ReferrerSource, "Hacker News"},
		"utm_source": {row.UTMSource, "hn"},
		"country":    {row.CountryCode, "DE"},
		"browser":    {row.Browser, "Chrome"},
		"device":     {row.Device, DeviceDesktop},
	}
	for name, c := range checks {
		if c.got != c.want {
			t.Errorf("%s = %q, want %q", name, c.got, c.want)
		}
	}
	if row.SiteID != 42 {
		t.Errorf("site_id = %d, want 42", row.SiteID)
	}
}

func TestBuildRow_NoPII(t *testing.T) {
	row := BuildRow(testBeacon(), testCtx("day-1"))
	// The IP and raw UA must never survive onto the stored row.
	fields := strings.Join([]string{
		row.Hostname, row.Pathname, row.Referrer, row.ReferrerSource,
		row.Browser, row.BrowserVersion, row.OS, row.OSVersion, row.Device,
		row.CountryCode, row.Region, row.City,
	}, " ")
	if strings.Contains(fields, "203.0.113.7") {
		t.Error("stored row contains the client IP")
	}
	if strings.Contains(fields, "Mozilla") {
		t.Error("stored row contains the raw User-Agent")
	}
}

func TestBuildRow_VisitorHashRotates(t *testing.T) {
	a := BuildRow(testBeacon(), testCtx("day-1"))
	b := BuildRow(testBeacon(), testCtx("day-1"))
	c := BuildRow(testBeacon(), testCtx("day-2"))
	if a.VisitorID != b.VisitorID {
		t.Error("same salt should produce a stable visitor hash")
	}
	if a.VisitorID == c.VisitorID {
		t.Error("different salt (day) should produce a different visitor hash")
	}
	if a.VisitorID == 0 {
		t.Error("visitor hash should be non-zero")
	}
}

func TestIsBot(t *testing.T) {
	if !IsBot("") {
		t.Error("empty UA should be treated as a bot")
	}
	if !IsBot("Googlebot/2.1 (+http://www.google.com/bot.html)") {
		t.Error("Googlebot should be detected")
	}
	if IsBot(chromeUA) {
		t.Error("Chrome should not be a bot")
	}
}

File: apps/ingestor/src/lib/pipeline.test.ts
import { describe, expect, it } from 'bun:test';
import { buildEventRow, type RequestContext } from './pipeline.js';
import type { Beacon } from '@seyr/shared';

const beacon: Beacon = {
	n: 'pageview',
	u: 'https://example.com/pricing?utm_source=hn',
	d: 'example.com',
	r: 'https://news.ycombinator.com/'
};

function ctx(overrides: Partial<RequestContext> = {}): RequestContext {
	return {
		headers: new Headers({ 'x-forwarded-for': '203.0.113.7', 'cf-ipcountry': 'DE' }),
		userAgent:
			'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120 Safari/537.36',
		salt: 'salt-day-1',
		siteId: 42n,
		...overrides
	};
}

describe('buildEventRow', () => {
	it('derives page, referrer, utm, geo, and device columns', () => {
		const row = buildEventRow(beacon, ctx());
		expect(row.site_id).toBe(42n);
		expect(row.hostname).toBe('example.com');
		expect(row.pathname).toBe('/pricing');
		expect(row.referrer_source).toBe('Hacker News');
		expect(row.utm_source).toBe('hn');
		expect(row.country_code).toBe('DE');
		expect(row.browser).toBe('Chrome');
		expect(row.device).toBe('desktop');
	});

	it('stores no PII — IP and raw UA never appear on the row', () => {
		const row = buildEventRow(beacon, ctx());
		const serialized = JSON.stringify(row, (_k, v) => (typeof v === 'bigint' ? v.toString() : v));
		expect(serialized).not.toContain('203.0.113.7');
		expect(serialized).not.toContain('Mozilla');
	});

	it('produces a stable visitor hash within a day, different across salts', () => {
		const a = buildEventRow(beacon, ctx({ salt: 'day-1' }));
		const b = buildEventRow(beacon, ctx({ salt: 'day-1' }));
		const c = buildEventRow(beacon, ctx({ salt: 'day-2' }));
		expect(a.visitor_id).toBe(b.visitor_id);
		expect(a.visitor_id).not.toBe(c.visitor_id);
	});
});
